src/unauthenticated_stream/mod.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! Unauthenticated stream cipher.
+//!
+//! # Warning
+//! This API does not provide any authentification! You should use the
+//! [aead](aead/index.html) module instead.
+//!
+//! C analog: `EVP_CIPHER`
+//!
+//! Go analog: [`crypto.cipher.Stream`]
+//!
+//! [`crypto.cipher.Stream`]: https://golang.org/pkg/crypto/cipher/#Stream
+
+use {chacha, error, init, polyfill};
+
+pub use self::chacha20::CHACHA20;
+
+/// A key for decrypting data encrypted using an unauthenticated stream cipher.
+///
+/// C analog: `EVP_CIPHER_CTX`
+///
+/// Go analog: [`crypto.cipher.Stream`]
+pub struct DecryptingKey {
+    key: Key,
+}
+
+impl DecryptingKey {
+    /// Create a new decrypting key.
+    ///
+    /// `key_bytes` must be exactly `algorithm.key_len` bytes long.
+    ///
+    /// C analogs: `EVP_CIPHER_CTX_init`.
+    ///
+    /// Go analog:
+    ///   [`crypto.aes.NewCipher`](https://golang.org/pkg/crypto/aes/#NewCipher)
+    #[inline]
+    pub fn new(algorithm: &'static Algorithm, key_bytes: &[u8])
+               -> Result<DecryptingKey, error::Unspecified> {
+        Ok(DecryptingKey {
+            key: Key::new(algorithm, key_bytes)?,
+        })
+    }
+
+    /// The key's unauthenticated stream cipher algorithm.
+    ///
+    /// C analog: `EVP_CIPHER_CTX.cipher`
+    #[inline(always)]
+    pub fn algorithm(&self) -> &'static Algorithm { self.key.algorithm() }
+}
+
+/// Decrypts the input data in place.
+pub fn decrypt_in_place<'a>(key: &DecryptingKey, nonce: &[u8],
+                            in_out: &'a mut [u8])
+                            -> Result<&'a mut [u8], error::Unspecified> {
+    let nonce = slice_as_array_ref!(nonce, NONCE_LEN)?;
+    check_per_nonce_max_bytes(key.key.algorithm, in_out.len())?;
```

Comment:
Shouldn't `xor_keystream` do the above two steps, since they are the same for encryption and decryption?

Reply:
Done (I did the "do_encryption_in_place" thing, so AES won't need changes).